store/iterator.go

```diff
@@ -108,163 +95,158 @@ func (b *btreeIter) wrap(parent Iterator) *itemIter {
 		wrap:   b,
 		parent: parent,
 	}
-	if err := iter.skipAllDeleted(); err != nil {
-		panic(err)
-	}
 	return iter
 }
 
-func (b *btreeIter) next() {
-	b.data, b.hasMore = <-b.read
+func (b *btreeIter) Next() (keyer, error) {
+	data, hasMore := <-b.read
+	if !hasMore {
+		return nil, errors.Wrap(errors.ErrDone, "btree iterator")
+	}
+	key, ok := data.(keyer)
+	if !ok {
+		return nil, errors.Wrapf(errors.ErrType, "expected keyer, got %T", data)
+	}
+	return key, nil
 }
 
-func (b *btreeIter) close() {
+func (b *btreeIter) Release() {
 	b.once.Do(func() {
 		b.stop <- struct{}{}
 	})
 }
 
-// get requires this is valid, gets what we are pointing at
-func (b *btreeIter) get() keyer {
-	return b.data.(keyer)
-}
-
-func (b *btreeIter) valid() bool {
-	return b.hasMore
-}
-
 type itemIter struct {
 	wrap *btreeIter
 	// if we are iterating in a cache-wrap (and who isn't),
 	// we need to combine this iterator with the parent
 	parent Iterator
+
+	parentDone   bool
+	cachedParent Model
+	wrapDone     bool
+	cachedWrap   keyer
 }
 
 //------- public facing interface ------
 var _ Iterator = (*itemIter)(nil)
 
-// Valid implements Iterator and returns true iff it can be read
-func (i *itemIter) Valid() bool {
-	return i.wrap.valid() || i.parentValid()
-}
-
-// Next moves the iterator to the next sequential key in the database, as
-// defined by order of iteration.
+// advanceParent will read next from parent iterators,
+// and set cached value as well as done flags.
 //
-// If Valid returns false, this method will panic.
-func (i *itemIter) Next() error {
-	// advance either us, parent, or both
-	switch i.firstKey() {
-	case us:
-		i.wrap.next()
-	case both:
-		i.wrap.next()
-		fallthrough
-	case parent:
-		err := i.parent.Next()
-		if err != nil {
-			return err
+// it will skip closed and missing iterators.
+// doesn't return ErrDone, but only unexpected data errors.
+func (i *itemIter) advanceParent() error {
+	if i.parent == nil {
+		i.parentDone = true
```

Comment:
pp: you can `return nil` in L142 already
The guard can also be extended in this case to `if i.parent == nil || i.parentDone {`

Reply:
simplified